### PR TITLE
Allow auto-enable and auto-start for deb-systemd services after install

### PR DESCRIPTION
**Pull request by c-ameron**
Hi, this PR adds in two new options to enable/disable and autostart systemd services

`--[no-]deb-systemd-enable` to enable the systemd service after install
`--[no-]deb-systemd-auto-start` to auto start the service after install

Additionally, any service start/stop/restart will use the `deb-systemd-invoke` wrapper if installed on the system

This should solve https://github.com/jordansissel/fpm/issues/1363 and https://github.com/jordansissel/fpm/issues/1284

Even though you can now disable a service from starting by default, I think its important to still allow for using [deb-systemd-invoke](http://manpages.ubuntu.com/manpages/trusty/man1/deb-systemd-invoke.1p.html) to keep in line with debian standards.

Let me know if you need anything changed! Thanks :)

### CODE REVIEW

**Comment by c-aamm** [comment]
I've updated this PR to fix https://github.com/jordansissel/fpm/issues/1271

We pass an array of service names to the post install script and start them individually instead of only the last service in  `deb_systemd_list`. Let me know if this implementation fits the style with the rest of fpm 😃

**Comment by ilederman** [comment]
I came across this today and it would be a great feature to have

**Comment by c-aamm** [comment]
Hi @jordansissel - any chance this could please be merged? I've just updated it to properly remove any services that get installed. Thanks!

**Comment by jordansissel** [comment]
I will review and test soon. thank you for making fpm better :)

**Comment by ilederman** [comment]
Hi, any updates to this?

**Comment by jordansissel** [comment]
Noted! I will merge this before the next release.

File: templates/deb/postinst_upgrade.sh.erb [hunk 1]
```diff
@@ -8,18 +8,26 @@ after_upgrade() {
 <%=  script(:after_upgrade) %>
 <% end -%>
 
-<% if attributes[:deb_systemd] -%>
+<%# if any systemd services specified, loop through and start them -%>
+<% unless attributes[:deb_systemd].empty? -%>
```
Comment (jordansissel) [inline templates/deb/postinst_upgrade.sh.erb]: For future style note, I prefer to avoid this kind of usage where a double-negative is used. I find it hard to parse (in my head). By double-negative I mean `unless` being a negative of `if` and `.empty?` being the negative of `.any?`. No change required at this time for this line.

"unless empty" can be written without negatives as "if any": `if attributes[:deb_systemd].any?`

In general, I prefer to avoid `unless` in most places with exception to an inline use `foo(..) unless condition` because I find `unless` a bit of a mental twist to read.

File: templates/deb/postinst_upgrade.sh.erb [hunk 2]
```diff
@@ -8,18 +8,26 @@ after_upgrade() {
...
 systemctl --system daemon-reload >/dev/null || true
-if ! systemctl is-enabled <%= attributes[:deb_systemd] %> >/dev/null 
+sysctl=$(command -v deb-systemd-invoke || echo systemctl)
+<% attributes[:deb_systemd].each do |service| -%>
+if ! systemctl is-enabled <%= service %> >/dev/null
```
Comment (jordansissel) [inline templates/deb/postinst_upgrade.sh.erb]: Did you mean `"$sysctl"` here instead of `systemctl` ? Also, for variable names, `sysctl` may confuse future maintainers as `sysctl` is a different tool from `systemctl`. This comment applies to any invocations of `systemctl` where `$sysctl` seems the appropriate replacement.

File: templates/deb/postinst_upgrade.sh.erb [hunk 3]
```diff
@@ -31,10 +39,18 @@ after_install() {
...
-<% if attributes[:deb_systemd] -%>
+<%# if any systemd services specified, loop through and start them -%>
+<% unless attributes[:deb_systemd].empty? -%>
 systemctl --system daemon-reload >/dev/null || true
-systemctl enable <%= attributes[:deb_systemd] %> >/dev/null || true
-systemctl start <%= attributes[:deb_systemd] %> >/dev/null || true
+sysctl=$(command -v deb-systemd-invoke || echo systemctl)
+<% attributes[:deb_systemd].each do |service| -%>
+<% if attributes[:deb_systemd_enable?]-%>
+systemctl enable <%= service %> >/dev/null || true
```
Comment (jordansissel) [inline templates/deb/postinst_upgrade.sh.erb]: did you mean `$sysctl` instead of `systemctl` here?

File: templates/deb/postinst_upgrade.sh.erb [hunk 4]
```diff
@@ -8,18 +8,26 @@ after_upgrade() {
...
 systemctl --system daemon-reload >/dev/null || true
-if ! systemctl is-enabled <%= attributes[:deb_systemd] %> >/dev/null 
+sysctl=$(command -v deb-systemd-invoke || echo systemctl)
+<% attributes[:deb_systemd].each do |service| -%>
+if ! systemctl is-enabled <%= service %> >/dev/null 
 then
-    systemctl enable <%= attributes[:deb_systemd] %> >/dev/null || true
-    systemctl start <%= attributes[:deb_systemd] %> >/dev/null || true
+<% if attributes[:deb_systemd_enable?]-%>
+    systemctl enable <%= service %> >/dev/null || true
```
Comment (jordansissel) [inline templates/deb/postinst_upgrade.sh.erb]: did you mean `$sysctl` instead of `systemctl` here?

File: templates/deb/prerm_upgrade.sh.erb [hunk 5]
```diff
@@ -8,10 +8,12 @@ before_remove() {
...
 <% end -%>
 
-<%# Removing the systemd service-%>
-<% if attributes[:deb_systemd] -%>
-systemctl stop <%= attributes[:deb_systemd] %> >/dev/null || true
-systemctl disable <%= attributes[:deb_systemd] %> >/dev/null || true
+<%# Stop and remove any systemd services that were installed-%>
+<% unless attributes[:deb_systemd].empty? -%>
+<% attributes[:deb_systemd].each do |service| -%>
+systemctl stop <%= service %> >/dev/null || true
```
Comment (jordansissel) [inline templates/deb/prerm_upgrade.sh.erb]: did you mean `$sysctl` instead of `systemctl` here?

File: templates/deb/prerm_upgrade.sh.erb [hunk 6]
```diff
@@ -8,10 +8,12 @@ before_remove() {
 <%=  script(:before_remove) %>
 <% end -%>
 
-<%# Removing the systemd service-%>
-<% if attributes[:deb_systemd] -%>
-systemctl stop <%= attributes[:deb_systemd] %> >/dev/null || true
-systemctl disable <%= attributes[:deb_systemd] %> >/dev/null || true
+<%# Stop and remove any systemd services that were installed-%>
+<% unless attributes[:deb_systemd].empty? -%>
+<% attributes[:deb_systemd].each do |service| -%>
```
Comment (jordansissel) [inline templates/deb/prerm_upgrade.sh.erb]: Did you want to use the same "try deb-systemd-invoke or systemctl" check here as your other scripts do?

File: lib/fpm/package/deb.rb [hunk 7]
```diff
@@ -428,26 +432,27 @@ def output(output_path)
...
       File.chmod(0644, dest_systemd)
 
-      # set the attribute with the systemd service name
-      attributes[:deb_systemd] = name
+      # add systemd service name to attribute
+      attributes[:deb_systemd] << name
     end
 
-    if script?(:before_upgrade) or script?(:after_upgrade) or attributes[:deb_systemd]
+    if script?(:before_upgrade) or script?(:after_upgrade) or not attributes[:deb_systemd].empty?
```
Comment (jordansissel) [inline lib/fpm/package/deb.rb]: Similar style comment (no change required) - double negatives can be hard to read, and `or attributes[...].any?` to me is clearer than `not attributes[...].empty?`

File: lib/fpm/package/deb.rb [hunk 8]
```diff
@@ -181,6 +181,10 @@ class FPM::Package::Deb < FPM::Package
     next File.expand_path(file)
   end
 
+  option "--systemd-enable", :flag , "Enable service on install or upgrade", :default => true
+
+  option "--systemd-auto-start", :flag , "Start service after install or upgrade", :default => true
```
Comment (jordansissel) [inline lib/fpm/package/deb.rb]: I would prefer this default to false. Otherwise, we're breaking behavior from a previous release where services were not automatically started. Users wishing to have a service auto start should be required to use this flag.

File: lib/fpm/package/deb.rb [hunk 9]
```diff
@@ -181,6 +181,10 @@ class FPM::Package::Deb < FPM::Package
     next File.expand_path(file)
   end
 
+  option "--systemd-enable", :flag , "Enable service on install or upgrade", :default => true
```
Comment (jordansissel) [inline lib/fpm/package/deb.rb]: I would prefer this default to false. Otherwise, we're breaking behavior from a previous release where services were not automatically started. Users wishing to have a service enabled out of the box should be required to use this flag.

File: templates/deb/postinst_upgrade.sh.erb (hunk 3 above)
Comment (c-ameron) [reply templates/deb/postinst_upgrade.sh.erb]: nope, `deb-systemd-invoke` doesn't have an `enable` option, so `systemctl` is fine to be used http://manpages.ubuntu.com/manpages/bionic/man1/deb-systemd-invoke.1p.html

File: templates/deb/postinst_upgrade.sh.erb (hunk 4 above)
Comment (c-ameron) [reply templates/deb/postinst_upgrade.sh.erb]: same comment as above

File: templates/deb/postinst_upgrade.sh.erb (hunk 2 above)
Comment (c-ameron) [reply templates/deb/postinst_upgrade.sh.erb]: good find, I've changed it `debsystemd` instead

**Comment by c-ameron** [comment]
@jordansissel Thanks for your help with the PR, sorry its taken a while to get onto this.
I've updated the code per your comments

- The enabling and autostart is off by default
- Use `debsystemd` as a command wrapper for _deb-systemd-invoke or systemctl_ instead of `$sysctl`
- Style changes per your comments

Please let me know if there's anything else I can do 😄

File: lib/fpm/package/deb.rb (hunk 8 above)
Comment (c-ameron) [reply lib/fpm/package/deb.rb]: I've defaulted both to false

File: templates/deb/postinst_upgrade.sh.erb (hunk 1 above)
Comment (c-ameron) [reply templates/deb/postinst_upgrade.sh.erb]: I've changed this to suit your style :)

**Comment by jordansissel** [comment]
> sorry its taken a while to get onto this

No worries. There's only so many hours in a day. :)

**Comment by vutny** [comment]
Awesome! Thank you @c-ameron and @jordansissel .
